R/simulate.r

```diff
@@ -1,3 +1,309 @@
+#' 4 (2): 279–95. \doi{https://doi.org/10.1093/biostatistics/4.2.279}.
+simulate_chains <- function(
+    index_cases,
+    statistic = c("size", "length"),
+    offspring_dist,
+    ...,
+    stat_max = Inf,
+    pop = Inf,
+    percent_immune = 0,
+    generation_time = NULL,
+    t0 = 0,
+    tf = Inf) {
+  # Input checking
+  checkmate::assert_count(index_cases, positive = TRUE)
+  checkmate::assert_choice(statistic, choices = c("size", "length"))
+  checkmate::assert_string(offspring_dist)
+  # check that offspring function exists in the environment
+  roffspring_name <- paste0("r", offspring_dist)
+  check_offspring_func_valid(roffspring_name)
+  checkmate::assert(
+    is.infinite(stat_max) ||
+      checkmate::assert_integerish(stat_max, lower = 0)
+  )
+  checkmate::assert(
+    is.infinite(pop) ||
+      checkmate::assert_integerish(pop, lower = 0)
+  )
+  checkmate::assert_number(
+    percent_immune,
+    lower = 0, upper = 1
+  )
+  if (!missing(generation_time)) {
+    check_generation_time_valid(generation_time)
+  } else if (!missing(tf)) {
+    stop("If `tf` is specified, `generation_time` must be specified too.")
+  }
+  checkmate::assert_numeric(
+    t0,
+    lower = 0, finite = TRUE
+  )
+  checkmate::assert_number(
+    tf,
+    lower = 0
+  )
+  # Gather offspring distribution parameters
+  pars <- list(...)
+
+  # Initialisations
+  stat_track <- rep(1, index_cases) # track statistic (length or size)
+  n_offspring <- rep(1, index_cases) # current number of offspring
+  sim <- seq_len(index_cases) # track trees that are still being simulated
+  infector_ids <- rep(1, index_cases)
+
+  # initialise list of data frame to hold the transmission trees
+  generation <- 1L
+  tree_df <- list(
+    data.frame(
+      infectee_id = seq_len(index_cases),
+      sim_id = 1L,
+      infector_id = NA_integer_,
+      generation = generation
+    )
+  )
+  # Initialise susceptible population
+  # I think this currently implicitly assumes the initial cases are imported
+  # (not subtracted from pop). If not, we'll have to subtract the non-imported
+  # cases. An improvement of this could take the form of an extra argument that
+  # defines a percentage of initial cases suspected to be imported. We can
+  # then substract only the non-imported cases from pop.
+  susc_pop <- round(pop * (1 - percent_immune))
+
+  # Add optional columns
+  if (!missing(generation_time)) {
+    tree_df[[generation]]$time <- t0
+    times <- tree_df[[generation]]$time
+  }
+  if (!missing(pop)) {
+    tree_df[[generation]]$susc_pop <- susc_pop
+  }
+  # next, simulate n trees
+  while (length(sim) > 0 && susc_pop > 0) {
+    # simulate next generation
+    next_gen <- do.call(
+      roffspring_name,
+      c(
+        list(n = sum(n_offspring[sim])),
+        pars
+      )
+    )
+    if (any(next_gen %% 1 > 0)) {
+      stop("Offspring distribution must return integers")
+    }
+    # Adjust next_gen if sum exceeds the number of susceptibles
```

Comment:
Do we not first need to adjust for the fact that some of the offspring won't be infected because they are no longer susceptible? I.e. something that does the same as this:
https://github.com/epiverse-trace/epichains/blob/8266e7c4e97e54a4f813a7c82a40311bf188bc37/R/simulate.r#L465
I think it could be achieved with:
```suggestion
    # Sample susceptible offspring to be infected from all possible offspring
    next_gen <- rbinom(n = next_gen, size = next_gen, prob = susc_pop / pop)
    # Adjust next_gen if sum exceeds the number of susceptibles
```

Reply:
@sbfnk I've made all the changes except this step that I don't quite understand. 

As `susc_pop/pop` approaches very small values (because `susc_pop` is running out), this will lead to 0 `next_gen` vectors that will cause the simulation to end even though the remaining susceptibles could have been infected. I believe that's why we do the next step of sampling true infections from the possible infections where `sum(next_gen) > susc_pop`. I am missing something here.

Reply:
> As susc_pop/pop approaches very small values (because susc_pop is running out), this will lead to 0 next_gen vectors that will cause the simulation to end even though the remaining susceptibles could have been infected

I think that's desired behaviour (stochastic fadeout due to herd immunity). I see it this way: before this line we sample `next_gen` from the offspring distribution. Here we check which of them are still susceptible (otherwise they can't be infected). If none of them are (with an element of chance) the infection dies out and nobody else gets infected.

Reply:
Awesome. That makes sense. Thanks.